File: parsing-api/src/services/llm.service.ts
import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';
import axios from 'axios';
import config from '../config/index.js';
import logger from '../utils/logger.js';
import type { ClassificationResult, TOON, ResumeTOON, JDTOON } from '../../../shared/types/toon.js';

/**
 * Initialize LLM clients based on configuration
 */
const openaiClient = config.llm.openai.apiKey 
  ? new OpenAI({ apiKey: config.llm.openai.apiKey })
  : null;

const anthropicClient = config.llm.anthropic.apiKey
  ? new Anthropic({ apiKey: config.llm.anthropic.apiKey })
  : null;

/**
 * Classify document type using LLM
 */
export async function classifyDocument(text: string): Promise<ClassificationResult> {
  const startTime = Date.now();
  const prompt = `Analyze the following document text and classify it as either a "resume" or "job_description".
Return ONLY a JSON object with this exact format:
{
  "type": "resume" | "job_description" | "unknown",
  "confidence": 0.0 to 1.0
}

Document text:
${text.substring(0, 2000)}`;

  try {
    let response: ClassificationResult;
    
    switch (config.llm.provider) {
      case 'openai':
        response = await classifyWithOpenAI(prompt);
        break;
      case 'anthropic':
        response = await classifyWithAnthropic(prompt);
        break;
      case 'xai':
        response = await classifyWithXAI(prompt);
        break;
      default:
        throw new Error(`Unknown LLM provider: ${config.llm.provider}`);
    }
    
    logger.info('Document classified', {
      type: response.type,
      confidence: response.confidence,
      duration: Date.now() - startTime,
    });
    
    return response;
  } catch (error) {
    logger.error('Classification failed', { error: error instanceof Error ? error.message : 'Unknown error' });
    throw error;
  }
}

/**
 * Parse document into TOON format using LLM
 */
export async function parseToTOON(text: string, documentType: 'resume' | 'job_description'): Promise<TOON> {
  const startTime = Date.now();
  const prompt = documentType === 'resume' 
    ? getResumeParsingPrompt(text)
    : getJDParsingPrompt(text);
  
  try {
    let toon: TOON;
    
    switch (config.llm.provider) {
      case 'openai':
        toon = await parseWithOpenAI(prompt, documentType);
        break;
      case 'anthropic':
        toon = await parseWithAnthropic(prompt, documentType);
        break;
      case 'xai':
        toon = await parseWithXAI(prompt, documentType);
        break;
      default:
        throw new Error(`Unknown LLM provider: ${config.llm.provider}`);
    }
    
    // Post-process: Extract URLs from raw text if LLM missed them (for resumes only)
    if (documentType === 'resume' && toon && 'person' in toon) {
      toon = extractUrlsFromText(toon as ResumeTOON, text) as TOON;
    }
    
    logger.info('Document parsed to TOON', {
      type: documentType,
      duration: Date.now() - startTime,
      provider: config.llm.provider,
    });
    
    return toon;
  } catch (error) {
    logger.error('TOON parsing failed', { error: error instanceof Error ? error.message : 'Unknown error' });
    throw error;
  }
}

/**
 * OpenAI Classification
 */
async function classifyWithOpenAI(prompt: string): Promise<ClassificationResult> {
  if (!openaiClient) throw new Error('OpenAI client not configured');
  
  const response = await openaiClient.chat.completions.create({
    model: config.llm.openai.model,
    messages: [{ role: 'user', content: prompt }],
    temperature: 0.1,
    response_format: { type: 'json_object' },
  });
  
  const content = response.choices[0].message.content || '{}';
  return JSON.parse(content) as ClassificationResult;
}

/**
 * OpenAI Parsing
 */
async function parseWithOpenAI(prompt: string, documentType: string): Promise<TOON> {
  if (!openaiClient) throw new Error('OpenAI client not configured');
  
  const response = await openaiClient.chat.completions.create({
    model: config.llm.openai.model,
    messages: [{ role: 'user', content: prompt }],
    temperature: 0.1,
    response_format: { type: 'json_object' },
  });
  
  const content = response.choices[0].message.content || '{}';
  const parsed = JSON.parse(content);
  
  return {
    ...parsed,
    type: documentType,
  } as TOON;
}

/**
 * Anthropic Classification
 */
async function classifyWithAnthropic(prompt: string): Promise<ClassificationResult> {
  if (!anthropicClient) throw new Error('Anthropic client not configured');
  
  const response = await anthropicClient.messages.create({
    model: config.llm.anthropic.model,
    max_tokens: 1024,
    messages: [{ role: 'user', content: prompt }],
  });
  
  const content = response.content[0];
  if (content.type !== 'text') throw new Error('Unexpected response type');
  
  // Extract JSON from response
  const jsonMatch = content.text.match(/\{[\s\S]*\}/);
  if (!jsonMatch) throw new Error('No JSON found in response');
  
  return JSON.parse(jsonMatch[0]) as ClassificationResult;
}

/**
 * Anthropic Parsing
 */
async function parseWithAnthropic(prompt: string, documentType: string): Promise<TOON> {
  if (!anthropicClient) throw new Error('Anthropic client not configured');
  
  const response = await anthropicClient.messages.create({
    model: config.llm.anthropic.model,
    max_tokens: 4096,
    messages: [{ role: 'user', content: prompt }],
  });
  
  const content = response.content[0];
  if (content.type !== 'text') throw new Error('Unexpected response type');
  
  // Extract JSON from response
  const jsonMatch = content.text.match(/\{[\s\S]*\}/);
  if (!jsonMatch) throw new Error('No JSON found in response');
  
  const parsed = JSON.parse(jsonMatch[0]);
  
  return {
    ...parsed,
    type: documentType,
  } as TOON;
}

/**
 * X.AI Grok Classification
 */
async function classifyWithXAI(prompt: string): Promise<ClassificationResult> {
  const response = await axios.post(
    `${config.llm.xai.baseUrl}/chat/completions`,
    {
      model: config.llm.xai.model,
      messages: [{ role: 'user', content: prompt }],
      temperature: 0.1,
    },
    {
      headers: {
        'Authorization': `Bearer ${config.llm.xai.apiKey}`,
        'Content-Type': 'application/json',
      },
      timeout: config.timeouts.llmRequest,
    }
  );
  
  const content = response.data.choices[0].message.content;
  const jsonMatch = content.match(/\{[\s\S]*\}/);
  if (!jsonMatch) throw new Error('No JSON found in response');
  
  return JSON.parse(jsonMatch[0]) as ClassificationResult;
}

/**
 * X.AI Grok Parsing
 */
async function parseWithXAI(prompt: string, documentType: string): Promise<TOON> {
  const response = await axios.post(
    `${config.llm.xai.baseUrl}/chat/completions`,
    {
      model: config.llm.xai.model,
      messages: [{ role: 'user', content: prompt }],
      temperature: 0.1,
    },
    {
      headers: {
        'Authorization': `Bearer ${config.llm.xai.apiKey}`,
        'Content-Type': 'application/json',
      },
      timeout: config.timeouts.llmRequest,
    }
  );
  
  const content = response.data.choices[0].message.content;
  const jsonMatch = content.match(/\{[\s\S]*\}/);
  if (!jsonMatch) throw new Error('No JSON found in response');
  
  const parsed = JSON.parse(jsonMatch[0]);
  
  return {
    ...parsed,
    type: documentType,
  } as TOON;
}

/**
 * Extract URLs from text and add to TOON if missing
 */
function extractUrlsFromText(toon: ResumeTOON, rawText: string): ResumeTOON {
  if (!toon.person) {
    toon.person = { name: '', email: '', phone: '' };
  }
  
  // Comprehensive URL pattern
  const urlPattern = /(https?:\/\/[^\s<>"')]+|www\.[^\s<>"')]+|linkedin\.com\/[^\s<>"')]+|github\.com\/[^\s<>"')]+|twitter\.com\/[^\s<>"')]+|x\.com\/[^\s<>"')]+|[a-zA-Z0-9][a-zA-Z0-9-]*\.[a-zA-Z]{2,}[^\s<>"')]*)/gi;
  const foundUrls = rawText.match(urlPattern) || [];
  
  // Extract LinkedIn URLs
  if (!toon.person.linkedin) {
    const linkedinUrls = foundUrls.filter(url => /linkedin/i.test(url) || /linked\.in/i.test(url));
    if (linkedinUrls.length > 0) {
      let url = linkedinUrls[0].replace(/[.,;:]+$/, '');
      toon.person.linkedin = url.startsWith('http') ? url : `https://${url}`;
    }
  }
  
  // Extract GitHub URLs
  if (!toon.person.github) {
    const githubUrls = foundUrls.filter(url => /github/i.test(url));
    if (githubUrls.length > 0) {
      let url = githubUrls[0].replace(/[.,;:]+$/, '');
      toon.person.github = url.startsWith('http') ? url : `https://${url}`;
    }
  }
  
  // Extract Twitter URLs
  if (!toon.person.twitter) {
    const twitterUrls = foundUrls.filter(url => /twitter/i.test(url) || /x\.com/i.test(url));
    if (twitterUrls.length > 0) {
      let url = twitterUrls[0].replace(/[.,;:]+$/, '');
      toon.person.twitter = url.startsWith('http') ? url : `https://${url}`;
    }
  }
  
  // Extract portfolio/website URLs
  if (!toon.person.portfolio && !toon.person.website) {
    const excludedDomains = ['linkedin', 'github', 'twitter', 'x.com', 'gmail', 'yahoo', 'outlook', 'hotmail', 'email', 'mail', 'edu', 'ac.', '.gov'];
    const portfolioUrls = foundUrls.filter(url => {
      const urlLower = url.toLowerCase();
      return !excludedDomains.some(domain => urlLower.includes(domain)) && 
             url.includes('.') && 
             url.length > 5;
    });
    if (portfolioUrls.length > 0) {
      let url = portfolioUrls[0].replace(/[.,;:]+$/, '');
      toon.person.portfolio = url.startsWith('http') ? url : `https://${url}`;
    }
  }
  
  // Collect remaining URLs into otherUrls
  if (!toon.person.otherUrls) {
    toon.person.otherUrls = [];
  }
  
  const categorized = new Set<string>();
  if (toon.person.linkedin) categorized.add(toon.person.linkedin.toLowerCase());
  if (toon.person.github) categorized.add(toon.person.github.toLowerCase());
  if (toon.person.twitter) categorized.add(toon.person.twitter.toLowerCase());
  if (toon.person.portfolio) categorized.add(toon.person.portfolio.toLowerCase());
  if (toon.person.website) categorized.add(toon.person.website.toLowerCase());
  
  for (const url of foundUrls) {
    const urlClean = url.replace(/[.,;:]+$/, '');
    if (!categorized.has(urlClean.toLowerCase()) && !toon.person.otherUrls!.includes(urlClean)) {
      const urlFinal = urlClean.startsWith('http') ? urlClean : `https://${urlClean}`;
      toon.person.otherUrls!.push(urlFinal);
    }
  }
  
  return toon;
}

/**
 * Resume Parsing Prompt
 */
function getResumeParsingPrompt(text: string): string {
  return `Parse the following resume and extract information into this EXACT JSON format. Extract ALL URLs including LinkedIn, GitHub, portfolio, personal website, Twitter, and any other URLs found:

{
  "person": {
    "name": "Full Name",
    "email": "email@example.com",
    "phone": "+1234567890",
    "linkedin": "https://linkedin.com/in/username",
    "github": "https://github.com/username",
    "portfolio": "https://portfolio.example.com",
    "website": "https://personal-website.com",
    "twitter": "https://twitter.com/username",
    "otherUrls": ["https://other-url.com"]
  },
  "summary": "Professional summary or objective",
  "skills": ["skill1", "skill2", "skill3"],
  "experience": [
    {
      "title": "Job Title",
      "company": "Company Name",
      "from": "YYYY-MM",
      "to": "YYYY-MM or Present",
      "years": 2.5,
      "description": "Brief description of role and achievements"
    }
  ],
  "education": [
    {
      "degree": "Degree Name",
      "field": "Field of Study",
      "institution": "University Name",
      "year": "YYYY",
      "gpa": "3.8/4.0"
    }
  ],
  "certifications": ["Cert 1", "Cert 2"],
  "languages": ["English", "Spanish"],
  "total_experience_years": 5
}

Important:
- Extract ALL information accurately including ALL URLs (LinkedIn, GitHub, portfolio, website, Twitter, etc.)
- Extract URLs even if they are in different formats (e.g., linkedin.com/in/username or www.linkedin.com/in/username)
- Use "Present" for current positions
- Calculate years for experience entries
- Calculate total_experience_years as sum of all experience
- If information is missing, use empty string or empty array
- Return ONLY valid JSON, no additional text

Resume text:
${text}`;
}

/**
 * Job Description Parsing Prompt
 */
function getJDParsingPrompt(text: string): string {
  return `Parse the following job description and extract information into this EXACT JSON format:

{
  "title": "Job Title",
  "company": "Company Name",
  "location": "City, State/Country",
  "employment_type": "Full-time/Part-time/Contract",
  "min_experience_years": 3,
  "max_experience_years": 5,
  "skills": ["skill1", "skill2", "skill3"],
  "responsibilities": ["responsibility1", "responsibility2"],
  "keywords": ["keyword1", "keyword2"],
  "qualifications": ["qualification1", "qualification2"],
  "salary_range": "$80,000 - $120,000"
}

CRITICAL INSTRUCTIONS:
- Extract the COMPANY NAME from the job description - look for phrases like "Company:", "About [Company]", "We are [Company]", or company name in header/footer
- Extract ALL information accurately including company name
- For experience, parse "3-5 years" as min_experience_years: 3, max_experience_years: 5
- If only one number, set min_experience_years only
- Extract key technical skills separately from requirements
- Include all important keywords for matching
- If company name is not found, use empty string
- If information is missing, use empty string, 0, or empty array
- Return ONLY valid JSON, no additional text

Job Description text:
${text}`;
}

export default {
  classifyDocument,
  parseToTOON,
};

File: backend/llm_service.py
"""
LLM Service for Resume and JD Parsing. Output format is TOON (Token-Oriented Object Notation).
Supports X.AI Grok (multi-key rotation), OpenAI, and Anthropic
"""
import os
import time
import requests
from typing import Dict, Any, Literal, Optional

from toon import toon_loads_flex

# Reuse a session for connection keep-alive (faster repeated calls)
_session: Optional[requests.Session] = None

def _get_session() -> requests.Session:
    global _session
    if _session is None:
        _session = requests.Session()
    return _session

# Load configuration
LLM_PROVIDER = os.getenv('LLM_PROVIDER', 'xai')  # xai, openai, anthropic
XAI_MODEL = os.getenv('XAI_MODEL', 'grok-4-fast-reasoning')  # Fast model for parsing
LLM_REQUEST_TIMEOUT = int(os.getenv('LLM_REQUEST_TIMEOUT', '45'))  # Lower = faster fail; 45s suits grok-4-fast
LLM_MAX_INPUT_CHARS = int(os.getenv('LLM_MAX_INPUT_CHARS', '0'))  # 0 = no trim; set e.g. 18000 to speed very long docs
OPENAI_API_KEY = os.getenv('OPENAI_API_KEY', '')
ANTHROPIC_API_KEY = os.getenv('ANTHROPIC_API_KEY', '')


def call_llm(prompt: str, doc_type: Literal['resume', 'jd']) -> Dict[str, Any]:
    """
    Call LLM to parse document
    
    Args:
        prompt: The prompt containing document text
        doc_type: Type of document ('resume' or 'jd')
    
    Returns:
        Parsed TOON format as dict
    """
    if LLM_MAX_INPUT_CHARS and len(prompt) > LLM_MAX_INPUT_CHARS:
        prompt = prompt[:LLM_MAX_INPUT_CHARS] + "\n\n[Document truncated for length. Extract from above.]"
    if LLM_PROVIDER == 'xai':
        return call_xai_grok(prompt, doc_type)
    elif LLM_PROVIDER == 'openai':
        return call_openai(prompt, doc_type)
    elif LLM_PROVIDER == 'anthropic':
        return call_anthropic(prompt, doc_type)
    else:
        raise ValueError(f"Unsupported LLM provider: {LLM_PROVIDER}")


def call_xai_grok(prompt: str, doc_type: str, service_id: str = "parsing") -> Dict[str, Any]:
    """Call X.AI Grok API with multi-key rotation and cooldown on failure. Optimized for grok-4-fast-reasoning."""
    from llm_key_manager import get_key_for_service, report_result

    url = "https://api.x.ai/v1/chat/completions"
    payload = {
        "model": XAI_MODEL,
        "messages": [
            {"role": "system", "content": get_system_prompt(doc_type)},
            {"role": "user", "content": prompt},
        ],
        "temperature": 0.2,
        "max_tokens": 2048,
    }
    keys_tried = 0
    last_error: Optional[str] = None
    key_count = 0
    try:
        mgr = __import__("llm_key_manager", fromlist=["KeyManager"]).KeyManager.get_instance()
        key_count = mgr._registry.count
    except Exception:
        key_count = 0
    max_keys_to_try = key_count if key_count else 1
    timeout = max(15, min(120, LLM_REQUEST_TIMEOUT))

    while keys_tried < max_keys_to_try:
        key_slot = get_key_for_service(service_id)
        if not key_slot:
            raise ValueError("No LLM API key configured (set HRMS_API_KEY_1..N or XAI_API_KEY)")
        slot_id, secret = key_slot
        keys_tried += 1
        headers = {
            "Authorization": f"Bearer {secret}",
            "Content-Type": "application/json",
        }
        t0 = time.perf_counter()
        try:
            session = _get_session()
            response = session.post(url, headers=headers, json=payload, timeout=timeout)
            latency_ms = (time.perf_counter() - t0) * 1000
            if response.ok:
                report_result(slot_id, True, response.status_code, latency_ms)
                result = response.json()
                content = result["choices"][0]["message"]["content"]
                return parse_llm_response(content)
            status = response.status_code
            report_result(slot_id, False, status, latency_ms)
            if 400 <= status < 500 and status != 429:
                raise ValueError(f"X.AI API error ({status})")
            last_error = f"X.AI API error ({status})"
            continue
        except requests.exceptions.Timeout:
            latency_ms = (time.perf_counter() - t0) * 1000
            report_result(slot_id, False, None, latency_ms)
            last_error = "X.AI API timeout"
            continue
        except requests.exceptions.RequestException as e:
            latency_ms = (time.perf_counter() - t0) * 1000
            report_result(slot_id, False, getattr(e, "response", None) and getattr(e.response, "status_code"), latency_ms)
            if hasattr(e, "response") and e.response is not None and 400 <= e.response.status_code < 500 and e.response.status_code != 429:
                raise ValueError("X.AI API request failed")
            last_error = "X.AI API network error"
            continue
    raise ValueError(last_error or "X.AI API call failed after all keys tried")


def call_openai(prompt: str, doc_type: str) -> Dict[str, Any]:
    """Call OpenAI API"""
    if not OPENAI_API_KEY:
        raise ValueError("OPENAI_API_KEY not configured")
    
    url = "https://api.openai.com/v1/chat/completions"
    headers = {
        "Authorization": f"Bearer {OPENAI_API_KEY}",
        "Content-Type": "application/json"
    }
    
    payload = {
        "model": "gpt-4-turbo-preview",
        "messages": [
            {
                "role": "system",
                "content": get_system_prompt(doc_type)
            },
            {
                "role": "user",
                "content": prompt
            }
        ],
        "temperature": 0.3,
        "max_tokens": 2000,
        "response_format": {"type": "json_object"}
    }
    
    # Retry logic
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = requests.post(url, headers=headers, json=payload, timeout=90)
            response.raise_for_status()
            
            result = response.json()
            content = result['choices'][0]['message']['content']
            
            return parse_llm_response(content)
            
        except (requests.exceptions.Timeout, requests.exceptions.RequestException) as e:
            if attempt < max_retries - 1:
                import time
                wait_time = (attempt + 1) * 2
                print(f"[LLM RETRY] OpenAI error, retrying in {wait_time}s...")
                time.sleep(wait_time)
                continue
            raise ValueError(f"OpenAI API error: {str(e)}")
    
    raise ValueError("OpenAI API call failed after all retries")


def call_anthropic(prompt: str, doc_type: str) -> Dict[str, Any]:
    """Call Anthropic Claude API"""
    if not ANTHROPIC_API_KEY:
        raise ValueError("ANTHROPIC_API_KEY not configured")
    
    url = "https://api.anthropic.com/v1/messages"
    headers = {
        "x-api-key": ANTHROPIC_API_KEY,
        "anthropic-version": "2023-06-01",
        "Content-Type": "application/json"
    }
    
    payload = {
        "model": "claude-3-sonnet-20240229",
        "max_tokens": 2000,
        "temperature": 0.3,
        "system": get_system_prompt(doc_type),
        "messages": [
            {
                "role": "user",
                "content": prompt
            }
        ]
    }
    
    # Retry logic
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = requests.post(url, headers=headers, json=payload, timeout=90)
            response.raise_for_status()
            
            result = response.json()
            content = result['content'][0]['text']
            
            return parse_llm_response(content)
            
        except (requests.exceptions.Timeout, requests.exceptions.RequestException) as e:
            if attempt < max_retries - 1:
                import time
                wait_time = (attempt + 1) * 2
                print(f"[LLM RETRY] Anthropic error, retrying in {wait_time}s...")
                time.sleep(wait_time)
                continue
            raise ValueError(f"Anthropic API error: {str(e)}")
    
    raise ValueError("Anthropic API call failed after all retries")


def get_system_prompt(doc_type: str) -> str:
    """Get system prompt for LLM. Required output format is TOON (Token-Oriented Object Notation)."""
    if doc_type == 'resume':
        return """You are an expert resume parser. Extract ALL information from the resume including EVERY URL. Return ONLY valid TOON (Token-Oriented Object Notation): one key-value per line, key: value, nested keys with dots, scalar lists with pipe. Example:

type: resume
person.name: Full Name
person.email: email@example.com
person.phone: +1234567890
person.location: City, State/Country
person.linkedin: https://linkedin.com/in/username
person.github: 
person.portfolio: 
person.website: 
person.twitter: 
person.otherUrls[0]:
summary: Professional summary
skills: skill1|skill2
experience.0.title: Job Title
experience.0.company: Company Name
experience.0.from: 2020-01
experience.0.to: 2023-12
experience.0.years: 3.9
education.0.degree: Bachelor of Science
education.0.field: Computer Science
education.0.institution: University Name
education.0.year: 2020
certifications: cert1|cert2
total_experience_years: 3.9

CRITICAL: Extract EVERY URL (LinkedIn, GitHub, portfolio, website, Twitter) into the person fields; use empty string if not found. Extract location/city/address (e.g. Mumbai, Bangalore, Delhi NCR, City - Country) into person.location. Use pipe (|) for lists of strings. Return ONLY the TOON block, no markdown, no explanations. You may also return valid JSON and it will be accepted."""
    
    else:  # jd
        return """You are an expert job description parser. Extract information and return ONLY valid TOON (Token-Oriented Object Notation): one key-value per line, key: value, lists with pipe. Example:

type: job_description
title: Job Title
company: Company Name
location: City, Country
salary_range: 50000-80000
min_experience_years: 2
max_experience_years: 5
skills: skill1|skill2
responsibilities: resp1|resp2
qualifications: qual1|qual2
keywords: keyword1|keyword2

CRITICAL: Extract the COMPANY NAME. Return ONLY the TOON block, no markdown. You may also return valid JSON and it will be accepted."""


def parse_llm_response(content: str) -> Dict[str, Any]:
    """Parse LLM response as TOON or legacy JSON into a dict."""
    parsed = toon_loads_flex(content)
    if not parsed:
        raise ValueError("Failed to parse LLM response as TOON or JSON")
    return parsed


def classify_document(text: str) -> Literal['resume', 'jd', 'unknown']:
    """
    Classify document type using simple heuristics
    
    Args:
        text: Document text
    
    Returns:
        Document type: 'resume', 'jd', or 'unknown'
    """
    text_lower = text.lower()
    
    # Resume indicators
    resume_keywords = ['resume', 'cv', 'curriculum vitae', 'experience', 'education', 'skills', 'objective']
    resume_score = sum(1 for kw in resume_keywords if kw in text_lower)
    
    # JD indicators
    jd_keywords = ['job description', 'responsibilities', 'requirements', 'qualifications', 'we are looking for', 'position']
    jd_score = sum(1 for kw in jd_keywords if kw in text_lower)
    
    if resume_score > jd_score and resume_score >= 2:
        return 'resume'
    elif jd_score > resume_score and jd_score >= 2:
        return 'jd'
    else:
        return 'unknown'

